### PR TITLE
[emails][m]: introduce handler for external modules - refs #26

### PR DESCRIPTION
**Pull request by zelima**
Note: this is still WIP (Eg tests are missing). Pushing to make sure we are aligned with how it should be implemented

### CODE REVIEW

File: auth/extensions.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+from importlib import import_module
+from importlib.util import find_spec
+
+import logging
+
+from .credentials import allowed_extensions
+
+
+def action_user(extension, user_info={}):
```
Comment (akariv) [inline auth/extensions.py]: This should perform the action on all installed extensions.

File: auth/extensions.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    if extension not in allowed_extensions:
```
Comment (akariv) [inline auth/extensions.py]: it's not "allowed_extensions" but "installed extensions" or "enabled extensions"

File: auth/credentials.py [hunk 3]
```diff
@@ -26,4 +26,16 @@
         lambda s: s.split(':', 1),
         allowed_services
     )
-)
\ No newline at end of file
+)
+
+# Allowed extentsions
```
Comment (akariv) [inline auth/credentials.py]: This is a list of all enabled extensions, so the word 'allowed' is wrong here.

Also 'alias' is not needed here.

File: auth/controllers.py [hunk 4]
```diff
@@ -149,6 +150,8 @@ def _get_token_from_profile(provider, profile, private_key):
     email = norm_profile['email']
     avatar_url = norm_profile['avatar_url']
     user = create_or_get_user(userid, name, username, email, avatar_url)
+    if user.get('new'):
+        action_user('datahub-emails', user)
```
Comment (akariv) [inline auth/controllers.py]: the whole point was to avoid having datahub-specific code in the core code-base.

File: auth/extensions.py (hunk 1 above)
Comment (akariv) [inline auth/extensions.py]: why 'action_user' always calls 'on_new_user'?

File: auth/extensions.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        return False
+
+    module_name = allowed_extensions[extension]
```
Comment (akariv) [inline auth/extensions.py]: split this logic to another function.

As a test, add another method here - 'on_login' which is called whenever a user logs in (regardless of new or old)

File: README.md [hunk 6]
```diff
@@ -19,16 +19,20 @@ A generic OAuth2 authentication service and user permission manager.
...
-- `DATABASE_URL`: A SQLAlchemy compatible database connection string (where user data is stored) 
+- `DATABASE_URL`: A SQLAlchemy compatible database connection string (where user data is stored)
 - `EXTERNAL_ADDRESS`: The hostname where this service is located on
-- `ALLOWED_SERVICES`: 
+- `ALLOWED_SERVICES`:
     Which permissions providers are available. A `;` delimited list of provider identifiers.
-    Each provider identifier takes the form of `[alias:]provider`, where `provider` is the name of a Python module 
+    Each provider identifier takes the form of `[alias:]provider`, where `provider` is the name of a Python module
     which exports a `get_permissions(service, userid)` function.
+- `ALLOWED_EXTENSIONS`:
```
Comment (akariv) [inline README.md]: see later comment re naming and aliases

File: auth/controllers.py (hunk 4 above)
Comment (zelima) [reply auth/controllers.py]: But we should call that new method form `auth/extensions.py` somewhere, no? How it's gonna work without that?

Comment (zelima) [reply auth/controllers.py]: Ok think got it - the problem is in `datahub-emails` not in the method being called, right?

**Pull request merged**